hcipy/coronagraphy/multi_scale.py

```diff
@@ -0,0 +1,158 @@
+            self.lyot_stop = lyot_stop
+        else:
+            self.lyot_stop = Apodizer(lyot_stop)
+
+        levels = int(np.ceil(np.log(q / 2) / np.log(scaling_factor))) + 1
+        qs = [2 * scaling_factor**i for i in range(levels)]
+        num_airys = [input_grid.shape / 2]
+
+        focal_grids = []
+        self.focal_masks = []
+        self.props = []
+
+        for i in range(1, levels):
+            num_airys.append(num_airys[i - 1] * window_size / (2 * qs[i - 1] * num_airys[i - 1]))
+
+        for i in range(levels):
+            q = qs[i]
+            num_airy = num_airys[i]
+
+            focal_grid = make_focal_grid(q, num_airy, pupil_diameter=pupil_diameter, reference_wavelength=1, focal_length=1)
+            focal_mask = complex_mask(focal_grid)
+
+            if i != levels - 1:
+                wx = windows.tukey(window_size, 1, False)
+                wy = windows.tukey(window_size, 1, False)
+                w = np.outer(wy, wx)
+
+                w = np.pad(w, (focal_grid.shape - w.shape) // 2, 'constant').ravel()
+                focal_mask *= 1 - w
+
+            for j in range(i):
+                fft = FastFourierTransform(focal_grids[j])
+                mft = MatrixFourierTransform(focal_grid, fft.output_grid)
+
+                focal_mask -= mft.backward(fft.forward(self.focal_masks[j]))
+
+            if i == 0:
+                prop = FourierFilter(input_grid, focal_mask, q)
+            else:
+                prop = FraunhoferPropagator(input_grid, focal_grid)
+
+            focal_grids.append(focal_grid)
+            self.focal_masks.append(focal_mask)
+            self.props.append(prop)
+
+    def forward(self, wavefront):
+        '''Propagate a wavefront through the coronagraph.
+
+        Parameters
+        ----------
+        wavefront : Wavefront
+            The wavefront to propagate. This wavefront is expected to be
+            in the pupil plane.
+
+        Returns
+        -------
+        Wavefront
+            The Lyot plane wavefront.
+        '''
+        wavelength = wavefront.wavelength
+        wavefront.wavelength = 1
+
+        for i, (mask, prop) in enumerate(zip(self.focal_masks, self.props)):
+            if i == 0:
+                lyot = Wavefront(prop.forward(wavefront.electric_field), input_stokes_vector=wavefront.input_stokes_vector)
+            else:
+                focal = prop(wavefront)
+                focal.electric_field *= mask
+                lyot.electric_field += prop.backward(focal).electric_field
+
+        lyot.wavelength = wavelength
+        wavefront.wavelength = wavelength
+
+        if self.lyot_stop is not None:
+            lyot = self.lyot_stop.forward(lyot)
+
+        return lyot
+
+    def backward(self, wavefront):
+        '''Propagate backwards through the coronagraph.
+
+        For a vortex coronagraph, this essentially a forward propagation through the same
+        coronagraph, but with the sign of its charge flipped.
```

Comment:
@syhaffert same here, except I find it less critical here so I am happy to move it or delete it. Please let me know!